presto-main/src/main/java/com/facebook/presto/execution/SqlStageExecution.java

```diff
@@ -440,6 +446,73 @@ private static Split createRemoteSplitFor(TaskId taskId, URI taskLocation)
     }
 
+    private synchronized void updateTaskStatus(TaskStatus taskStatus)
+    {
+        try {
+            // always update done tasks before, state transitions to ensure
+            // the transition to "final status info" is not missed
+            if (taskStatus.getState().isDone()) {
+                doneTasks.add(taskStatus.getTaskId());
+            }
+
+            StageState stageState = getState();
+            if (stageState.isDone()) {
+                return;
+            }
+
+            TaskState taskState = taskStatus.getState();
+            if (taskState == TaskState.FAILED) {
+                RuntimeException failure = taskStatus.getFailures().stream()
+                        .findFirst()
+                        .map(this::rewriteTransportFailure)
+                        .map(ExecutionFailureInfo::toException)
+                        .orElse(new PrestoException(GENERIC_INTERNAL_ERROR, "A task failed for an unknown reason"));
+                stateMachine.transitionToFailed(failure);
+            }
+            else if (taskState == TaskState.ABORTED) {
+                // A task should only be in the aborted state if the STAGE is done (ABORTED or FAILED)
+                stateMachine.transitionToFailed(new PrestoException(GENERIC_INTERNAL_ERROR, "A task is in the ABORTED state but stage is " + stageState));
+            }
+            else if (taskState == TaskState.FINISHED) {
+                finishedTasks.add(taskStatus.getTaskId());
+            }
+
+            if (stageState == StageState.SCHEDULED || stageState == StageState.RUNNING) {
+                if (taskState == TaskState.RUNNING) {
+                    stateMachine.transitionToRunning();
+                }
+                if (finishedTasks.containsAll(allTasks)) {
+                    stateMachine.transitionToFinished();
+                }
+            }
+        }
+        finally {
+            // after updating state, check if all tasks have final status information
+            if (stateMachine.getState().isDone() && doneTasks.containsAll(allTasks)) {
+                stateMachine.setAllTasksFinal();
+            }
+        }
+    }
+
+    private ExecutionFailureInfo rewriteTransportFailure(ExecutionFailureInfo executionFailureInfo)
+    {
+        if (executionFailureInfo.getRemoteHost() != null &&
+                failureDetector.getState(executionFailureInfo.getRemoteHost()) == GONE) {
+            return new ExecutionFailureInfo(
+                    executionFailureInfo.getType(),
+                    executionFailureInfo.getMessage(),
+                    executionFailureInfo.getCause(),
+                    executionFailureInfo.getSuppressed(),
+                    executionFailureInfo.getStack(),
+                    executionFailureInfo.getErrorLocation(),
+                    REMOTE_HOST_GONE.toErrorCode(),
+                    executionFailureInfo.getRemoteHost());
+        }
+        else {
```

Comment:
unnecessary else

Reply:
I just moved that code, but I'll fix it also.